File: singleton/java/SicSingleton.java
package io.github.androidhot.designpattern.singleton.java;

// Static Internal Class singleton pattern.
public class SicSingleton {

  private SicSingleton() {

  }

  public static SicSingleton getInstance() {
    return SingletonHolder.sInstance;
  }

  private static class SingletonHolder {
    private static final SicSingleton sInstance = new SicSingleton();
  }
}

File: singleton/kotlin/SicSingleton.kt
package io.github.androidhot.designpattern.singleton.kotlin

// Static Internal Class singleton pattern.
class SicSingleton private constructor() {
    companion object {
        fun getInstance(): SicSingleton {
            return Inner.instance
        }
    }

    private object Inner {
        val instance = SicSingleton()
    }
}
